### Ingest

Goal: Get all cards read in from JSON and saved as a pickle. Keep as much data as will be useful, drop the rest. 

In [1]:
import numpy as np
import pandas as pd
import json
import re

To download all the cards run:

In [7]:
# import all magic cards ever printed, from web 14mb

# allsets = pd.read_json('http://mtgjson.com/json/AllSetsArray.json')
# allsets.info()

Else run this to load a locally stored file (and be nice to our kind friend's api)

In [8]:
# import all magic cards from file 

allsets = pd.read_json('data/AllSetsArray.json')
allsets.tail(2)

block                                            booster  \
185  Battle for Zendikar  [[rare, mythic rare], uncommon, uncommon, unco...   
186                  NaN                                                NaN   

    border                                              cards code  \
185  black  [{u'layout': u'normal', u'name': u'Bane of Bal...  BFZ   
186  black  [{u'layout': u'normal', u'name': u'Prairie Str...  EXP   

    gathererCode magicCardsInfoCode magicRaritiesCodes                  name  \
185          NaN                bfz                NaN   Battle for Zendikar   
186          NaN                exp                NaN  Zendikar Expeditions   

    oldCode  onlineOnly releaseDate       type  
185     NaN         NaN  2015-10-02  expansion  
186     NaN         NaN  2015-10-02      promo


Top level dataframe is 186 rows of one set per row. Cards for that set are stored in the "cards" colum as another JSON array. Sets tend to have 200-350 cards in each.

In [9]:
# drop columns that are full of nan and/or junk

allsets = allsets.drop(['booster', 'gathererCode', 'magicCardsInfoCode', 
                        'magicCardsInfoCode', 'code',
                        'oldCode', 'onlineOnly', 'magicRaritiesCodes'
                        ], axis=1)

### Grab the most recent set and check it out

In [99]:
# zoom in on cards from latest set, 'Battle for Zendikar'

zen = pd.read_json(json.dumps(allsets.cards[185]))

# add set name and date
zen['set'] = allsets.name[185]
zen['releaseDate'] = allsets.releaseDate[185]

zen.head(2)

artist  cmc colors  \
0    Chase Stone    7    NaN   
1  Todd Lockwood    5    NaN   

                                              flavor  \
0  The continent of Bala Ged was ravaged by massi...   
1                                                NaN   

                                         id         imageName  layout  \
0  b2baccd791f5ada8bbd505f11bf81e70c8ea27ab  bane of bala ged  normal   
1  cfff5e44e0263f10015fcf613b0f40065acf00b7     blight herder  normal   

   loyalty manaCost  multiverseid     ...        rarity              subtypes  \
0      NaN      {7}        401814     ...      Uncommon             [Eldrazi]   
1      NaN      {5}        401819     ...          Rare  [Eldrazi, Processor]   

  supertypes                                               text toughness  \
0        NaN  Whenever Bane of Bala Ged attacks, defending p...         5   
1        NaN  When you cast Blight Herder, you may put two c...         5   

                           type       types variations                  set  \
0            Creature — Eldrazi  [Creature]        NaN  Battle for Zendikar   
1  Creature — Eldrazi Processor  [Creature]        NaN  Battle for Zendikar   

  releaseDate  
0  2015-10-02  
1  2015-10-02  

[2 rows x 23 columns]

In [100]:
# drop junk 
zen.drop(['id', 'layout', 'multiverseid', 'imageName', 'subtypes', 
          'supertypes', 'variations', 'loyalty', 'number'], 
         axis=1, inplace=True)

zen.head(2)

artist  cmc colors  \
0    Chase Stone    7    NaN   
1  Todd Lockwood    5    NaN   

                                              flavor manaCost  \
0  The continent of Bala Ged was ravaged by massi...      {7}   
1                                                NaN      {5}   

               name power    rarity  \
0  Bane of Bala Ged     7  Uncommon   
1     Blight Herder     4      Rare   

                                                text toughness  \
0  Whenever Bane of Bala Ged attacks, defending p...         5   
1  When you cast Blight Herder, you may put two c...         5   

                           type       types                  set releaseDate  
0            Creature — Eldrazi  [Creature]  Battle for Zendikar  2015-10-02  
1  Creature — Eldrazi Processor  [Creature]  Battle for Zendikar  2015-10-02

In [101]:
# grab the text from card 1

zen['text'][1]

u'When you cast Blight Herder, you may put two cards your opponents own from exile into their owners\' graveyards. If you do, put three 1/1 colorless Eldrazi Scion creature tokens onto the battlefield. They have "Sacrifice this creature: Add {1} to your mana pool."'

In [102]:
# fix errors

zen['text'].replace("\n" , " ", inplace=True, regex=True)
zen['text'].replace("\'" , "", inplace=True, regex=True)

zen['text'][1]

u'When you cast Blight Herder, you may put two cards your opponents own from exile into their owners graveyards. If you do, put three 1/1 colorless Eldrazi Scion creature tokens onto the battlefield. They have "Sacrifice this creature: Add {1} to your mana pool."'

### More cards

Now that we know everything words for one set, lets grab a bunch of sets and make one large table of cards. "Modern" is a major rules and format change the creators of the game enacted in 2003. It marks a turning point in the design, so, to start, we're only looking at cards after that date.  

In [10]:
# slice sets for all of modern (arbritrary rules change date in 2003)

modern_sets = allsets[ (allsets['releaseDate'] >= '2003-07-28' ) ]
modern_sets = modern_sets.loc[modern_sets['type'].isin(['core', 'expansion'])]
modern_sets = modern_sets.loc[modern_sets['border'].isin(['white', 'black'])]

modern_sets.reset_index(inplace=True)
modern_sets

index                block border  \
0      71                  NaN  white   
1      72             Mirrodin  black   
2      73             Mirrodin  black   
3      74             Mirrodin  black   
4      75             Kamigawa  black   
5      77             Kamigawa  black   
6      78             Kamigawa  black   
7      79                  NaN  white   
8      80              Ravnica  black   
9      83              Ravnica  black   
10     85              Ravnica  black   
11     86              Ice Age  black   
12     88          Time Spiral  black   
13     89          Time Spiral  black   
14     91          Time Spiral  black   
15     94          Time Spiral  black   
16     95                  NaN  black   
17     98               Lorwyn  black   
18    101               Lorwyn  black   
19    103           Shadowmoor  black   
20    105           Shadowmoor  black   
21    109                Alara  black   
22    111                Alara  black   
23    113                Alara  black   
24    114                  NaN  black   
25    118             Zendikar  black   
26    121             Zendikar  black   
27    123             Zendikar  black   
28    126                  NaN  black   
29    129    Scars of Mirrodin  black   
30    132    Scars of Mirrodin  black   
31    134    Scars of Mirrodin  black   
32    136                  NaN  black   
33    139            Innistrad  black   
34    141            Innistrad  black   
35    143            Innistrad  black   
36    145                  NaN  black   
37    148    Return to Ravnica  black   
38    150    Return to Ravnica  black   
39    153    Return to Ravnica  black   
40    155                  NaN  black   
41    158               Theros  black   
42    160               Theros  black   
43    162               Theros  black   
44    166                  NaN  black   
45    170      Khans of Tarkir  black   
46    177      Khans of Tarkir  black   
47    179      Khans of Tarkir  black   
48    182                  NaN  black   
49    185  Battle for Zendikar  black   

                                                cards  \
0   [{u'layout': u'normal', u'name': u'Abyssal Spe...   
1   [{u'layout': u'normal', u'name': u'Æther Spell...   
2   [{u'layout': u'normal', u'name': u'Æther Snap'...   
3   [{u'layout': u'normal', u'name': u'Abuna's Cha...   
4   [{u'layout': u'normal', u'name': u'Akki Avalan...   
5   [{u'layout': u'normal', u'name': u'Akki Blizza...   
6   [{u'layout': u'normal', u'name': u'Adamaro, Fi...   
7   [{u'layout': u'normal', u'name': u'Adarkar Was...   
8   [{u'layout': u'normal', u'name': u'Agrus Kos, ...   
9   [{u'layout': u'normal', u'name': u'Absolver Th...   
10  [{u'layout': u'normal', u'name': u'Æthermage's...   
11  [{u'layout': u'normal', u'name': u'Adarkar Val...   
12  [{u'layout': u'normal', u'name': u'Academy Rui...   
13  [{u'layout': u'normal', u'name': u'Akroma, Ang...   
14  [{u'layout': u'normal', u'name': u'Aeon Chroni...   
15  [{u'layout': u'normal', u'name': u'Akroma's Me...   
16  [{u'layout': u'normal', u'name': u'Abundance',...   
17  [{u'layout': u'normal', u'name': u'Adder-Staff...   
18  [{u'layout': u'normal', u'name': u'Ambassador ...   
19  [{u'layout': u'normal', u'name': u'Advice from...   
20  [{u'layout': u'normal', u'name': u'Aerie Ouphe...   
21  [{u'layout': u'normal', u'name': u'Ad Nauseam'...   
22  [{u'layout': u'normal', u'name': u'Absorb Vis'...   
23  [{u'layout': u'normal', u'name': u'Anathemance...   
24  [{u'layout': u'normal', u'name': u'Acidic Slim...   
25  [{u'layout': u'normal', u'name': u'Adventuring...   
26  [{u'layout': u'normal', u'name': u'Abyssal Per...   
27  [{u'layout': u'normal', u'name': u'Affa Guard ...   
28  [{u'layout': u'normal', u'name': u'Acidic Slim...   
29  [{u'layout': u'normal', u'name': u'Abuna Acoly...   
30  [{u'layout': u'normal', u'name': u'Accorder Pa...   
31  [{u'layout': u'normal', u'name': u'Act of Aggr...   
32  [{u'layout': u'normal', u'n

In [12]:
# groups cards from sets into one data frame 

cards = None

for s in xrange(len(modern_sets)):
    print "Reading in set:", modern_sets.name[s]
    target = pd.read_json(json.dumps(modern_sets.cards[s]))
    
    # slice to just the good stuff 
    target = target[['artist', 'cmc', 'colors', 'flavor', 'manaCost', 'name', 
                    'power', 'rarity', 'text', 'toughness', 'type', 'types']]
    
    # add set name and date
    target['set'] = modern_sets.name[s]
    target['releaseDate'] = modern_sets.releaseDate[s]    
    
    # add to cards df 
    cards = pd.concat([cards, target])

# clean up errors
cards['text'].replace("\n" , " ", inplace=True, regex=True)
cards['text'].replace("\'" , "", inplace=True, regex=True)
cards['flavor'].replace("\n" , " ", inplace=True, regex=True)

print 
print cards.info(verbose=False)

Reading in set: Eighth Edition
Reading in set: Mirrodin
Reading in set: Darksteel
Reading in set: Fifth Dawn
Reading in set: Champions of Kamigawa
Reading in set: Betrayers of Kamigawa
Reading in set: Saviors of Kamigawa
Reading in set: Ninth Edition
Reading in set: Ravnica: City of Guilds
Reading in set: Guildpact
Reading in set: Dissension
Reading in set: Coldsnap
Reading in set: Time Spiral
Reading in set: Time Spiral "Timeshifted"
Reading in set: Planar Chaos
Reading in set: Future Sight
Reading in set: Tenth Edition
Reading in set: Lorwyn
Reading in set: Morningtide
Reading in set: Shadowmoor
Reading in set: Eventide
Reading in set: Shards of Alara
Reading in set: Conflux
Reading in set: Alara Reborn
Reading in set: Magic 2010
Reading in set: Zendikar
Reading in set: Worldwake
Reading in set: Rise of the Eldrazi
Reading in set: Magic 2011
Reading in set: Scars of Mirrodin
Reading in set: Mirrodin Besieged
Reading in set: New Phyrexia
Reading in set: Magic 2012
Reading in set: Inni

This gives us 11502 cards

but this is too many cards, included are promotional tokens that we do not want. 

In [13]:
# filter out lands and tokens 

cards = cards.loc[cards['rarity'].isin(['Common', 'Uncommon', 'Rare', 
                                        'Mythic Rare'])]
cards.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10765 entries, 0 to 248
Columns: 14 entries, artist to releaseDate
dtypes: float64(1), object(13)
memory usage: 1.2+ MB


This gives 10765 cards.

Now the next problem is some cards do not have a color (artifacts) and some cards are multiple colors (gold, hybrid). Lets leave them out for now.  

In [14]:
# only keep cards with a 'color' attribute
cards_no_nulls = cards[cards['colors'].notnull()]

# only keep cards with text
cards_no_nulls = cards_no_nulls[cards_no_nulls['text'].notnull()]

# only keep cards with a mana cost
cards_no_nulls = cards_no_nulls[cards_no_nulls['colors'].map(len) == 1]
cards_no_nulls.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7874 entries, 0 to 197
Columns: 14 entries, artist to releaseDate
dtypes: float64(1), object(13)
memory usage: 922.7+ KB


In [15]:
# reset index and drop old index vals 

cards_no_nulls.reset_index(inplace=True)
cards_no_nulls.pop('index')
# cards_no_nulls.pop('level_0')
cards_no_nulls

artist  cmc   colors  \
0             Michael Sutfin    4  [Black]   
1              Wayne England    5   [Blue]   
2              Junko Taguchi    4  [Black]   
3     Alex Horley-Orlandelli    4    [Red]   
4          Melissa A. Benson    5  [White]   
5               Marc Fishman    2  [White]   
6            Donato Giancola    4   [Blue]   
7              Paolo Parente    5  [White]   
8                   Mark Zug    8  [White]   
9               Justin Sweet    4  [White]   
10       Christopher Moeller    4   [Blue]   
11    Greg & Tim Hildebrandt    5  [White]   
12                  Adam Rex    5   [Blue]   
13      Edward P. Beard, Jr.    1  [Green]   
14             Paolo Parente    3  [Green]   
15    Alex Horley-Orlandelli    1    [Red]   
16       Christopher Moeller    2  [White]   
17             Todd Lockwood    5  [White]   
18           Franz Vohwinkel    3    [Red]   
19               Daren Bader    6    [Red]   
20            Carl Critchlow    2  [Black]   
21               Dave Dorman    4  [Black]   
22    Jason Alexander Behnke    4    [Red]   
23           Alan Rabinowitz    2   [Blue]   
24           Andrew Robinson    5   [Blue]   
25             Paolo Parente    4  [Green]   
26          Christopher Rush    2  [Green]   
27                Gary Leach    2    [Red]   
28             Tony Szczudlo    3  [Black]   
29                     Berry    3   [Blue]   
...                      ...  ...      ...   
7844           Yeong-Hao Han    4    [Red]   
7845            Chris Rallis    4    [Red]   
7846         Anthony Palumbo    2  [Green]   
7847               Dan Scott    2  [Green]   
7848            Lake Hurwitz    4  [Green]   
7849          Eric Deschamps    6  [Green]   
7850        Daniel Ljunggren    4  [Green]   
7851                 Daarken    1  [Green]   
7852           Willian Murai    3  [Green]   
7853          Eric Deschamps    4  [Green]   
7854             Wesley Burt    3  [Green]   
7855           Lius Lasahido    6  [Green]   
7856              Chris Rahn    6  [Green]   
7857  Anastasia Ovchinnikova    2  [Green]   
7858           Jama Jurabaev    7  [Green]   
7859            Aaron Miller    2  [Green]   
7860      Bastien L. Deharme    4  [Green]   
7861           Kieran Yanner    3  [Green]   
7862           Yeong-Hao Han    2  [Green]   
7863        Daniel Ljunggren    1  [Green]   
7864         Anna Steinbauer    2  [Green]   
7865              Kev Walker    2  [Green]   
7866       Magali Villeneuve    2  [Green]   
7867        Daniel Ljunggren    2  [Green]   
7868          Greg Opalinski    6  [Green]   
7869             Wesley Burt    3  [Green]   
7870  Anastasia Ovchinnikova    5  [Green]   
7871            Matt Stewart    5  [Green]   
7872          Tyler Jacobson    3  [Green]   
7873           Vincent Proce    4  [Green]   

                                                 flavor      manaCost  \
0            To gaze under its hood is to invite death.     {2}{B}{B}   
1     Pray that it doesn't seek the safety of your l...     {3}{U}{U}   
2                        "Knowledge demands sacrifice."        {3}{B}   
3               Just try taking this bull by the horns.        {3}{R}   
4            A song of life soars over fields of blood.        {4}{W}   
5     If only every message were as perfect as its b...        {1}{W}   
6     "Words—so innocent and powerless are they, as ...     {2}{U}{U}   
7     Knights fight for honor and mercenaries fight ...        {4}{W}   
8                                                   NaN     {6}{W}{W}   
9                                                   NaN        {3}{W}   
10    The same spears that catch their food today wi...        {3}{U}   
11    Just as each added feather steadies the wing, ...        {4}{W}   
12    "To empty your mind is to fill your grave." —L...     {3}{U}{U}   
13                                                  NaN           {G}   
14    "There's never been a dwarven smith who could ...        {2}{G}   
1

In [16]:
# datamunge to get "color" out of a list format

def nolist(x):
    return x[0]

cards_no_nulls['colors'] = cards_no_nulls['colors'].apply(nolist)

Some cards have resource symbols in their text with their color. This might give a classifier an unfair advantage. 

In [17]:
# remove resource symbols from one card 

a = cards_no_nulls.text[3]
re.sub("{[A-Z]}" , "{1}", a)

u'{1}, {1}: Anaba Shaman deals 1 damage to target creature or player.'

In [18]:
# remove resource symbols from all cards 

def tap(x):
    return re.sub("{T}" , "Tap ", x)

def nomana(x):
    return re.sub("{[A-Z]}" , "{1}", x)

cards_no_nulls.text = cards_no_nulls.text.apply(tap)
cards_no_nulls.text = cards_no_nulls.text.apply(nomana)

In [19]:
cards_no_nulls.text

0       Flying Whenever Abyssal Specter deals damage t...
1                                                  Flying
2               You draw three cards and you lose 3 life.
3       {1}, Tap : Anaba Shaman deals 1 damage to targ...
4       Flying When Angel of Mercy enters the battlefi...
5       Flying Tap : Target attacking or blocking crea...
6                                      Tap : Draw a card.
7                                               Vigilance
8       If you have 3 or less life, Avatar of Hope cos...
9       Flying (This creature cant be blocked except b...
10      Flying (This creature cant be blocked except b...
11      Flying (This creature cant be blocked except b...
12      If target opponent has more cards in hand than...
13      Flying Tap : Add one mana of any color to your...
14      Enchant creature (Target a creature as you cas...
15      Blaze deals X damage to target creature or pla...
16       You gain 3 life for each creature attacking you.
17      Flying

In [20]:
# store to disk

# cards_no_nulls.reset_index(inplace=True)
# cards_no_nulls.pop('index')

cards_no_nulls.to_pickle('data/cards_modern.pkl')

Some cards have their name in the text, names have no predictive ability and could only lead to overfitting. 

In [21]:
# remove name from one card 

re.sub(cards_no_nulls.name[3] , "This", cards_no_nulls.text[3] )

u'{1}, Tap : This deals 1 damage to target creature or player.'

In [24]:
# remove card names from all cards  

import time

t0 = time.time()

for i in xrange(len(cards_no_nulls)):
    cards_no_nulls['text'][i] = re.sub(cards_no_nulls['name'][i] , "This", cards_no_nulls['text'][i]) 

t1 = time.time()

print round((t1-t0)/60, 2), "minutes"
cards_no_nulls.text

9.96 minutes


C:\Users\HN\Anaconda2\lib\site-packages\ipykernel\__main__.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


0       Flying Whenever This deals damage to a player,...
1                                                  Flying
2               You draw three cards and you lose 3 life.
3       {1}, Tap : This deals 1 damage to target creat...
4       Flying When This enters the battlefield, you g...
5       Flying Tap : Target attacking or blocking crea...
6                                      Tap : Draw a card.
7                                               Vigilance
8       If you have 3 or less life, This costs {6} les...
9       Flying (This creature cant be blocked except b...
10      Flying (This creature cant be blocked except b...
11      Flying (This creature cant be blocked except b...
12      If target opponent has more cards in hand than...
13      Flying Tap : Add one mana of any color to your...
14      Enchant creature (Target a creature as you cas...
15      This deals X damage to target creature or player.
16       You gain 3 life for each creature attacking you.
17      Flying

In [26]:
# store to disk

# cards_no_nulls.reset_index(inplace=True)
# cards_no_nulls.pop('index')

cards_no_nulls.to_pickle('data/cards_modern_no_name.pkl')

Some cards have parentheses explaining what a rules term means. All this text is redundanct and could only mislead a classifer. 

In [41]:
# remove helper text from one card 

a = cards_no_nulls.text[9]
re.sub('\([^)]*\)' , "", a)

u'Flying  When This enters the battlefield, destroy target enchantment.'

In [42]:
# remove helper text 

def hardmode(x):
    return re.sub('\([^)]*\)' , "", x)

cards_no_nulls.text = cards_no_nulls.text.apply(hardmode)

In [45]:
# store to disk

cards_no_nulls.to_pickle('data/5color_modern_no_name_hardmode.pkl')

### New DF with all modern cards



In [46]:
# groups all modern cards of all colors from sets into one data frame 

cards = None

for s in xrange(len(modern_sets)):
    print "Reading in set:", modern_sets.name[s]
    target = pd.read_json(json.dumps(modern_sets.cards[s]))
    
    # slice to just the good stuff 
    target = target[['artist', 'cmc', 'colors', 'flavor', 'manaCost', 'name', 
                    'power', 'rarity', 'text', 'toughness', 'type', 'types']]
    
    # add set name and date
    target['set'] = modern_sets.name[s]
    target['releaseDate'] = modern_sets.releaseDate[s]    
    
    # add to cards df 
    cards = pd.concat([cards, target])

# clean up errors
cards['text'].replace("\n" , " ", inplace=True, regex=True)
cards['text'].replace("\'" , "", inplace=True, regex=True)

print 
print cards.info(verbose=False)

# drop tokens 
cards_no_nulls = cards.loc[cards['rarity'].isin(['Common', 'Uncommon', 
                                                 'Rare', 'Mythic Rare'])]

# only keep cards with text
cards_no_nulls = cards_no_nulls[cards_no_nulls['text'].notnull()]

# save the "tap" symbol
cards_no_nulls.text = cards_no_nulls.text.apply(tap)

# flatten all resource symbols from all cards 
cards_no_nulls.text = cards_no_nulls.text.apply(nomana)

# remove helper text 
cards_no_nulls.text = cards_no_nulls.text.apply(hardmode)

# reset index
cards_no_nulls.reset_index(inplace=True)
cards_no_nulls.pop('index')

print "Done!"
cards_no_nulls.info()

Reading in set: Eighth Edition
Reading in set: Mirrodin
Reading in set: Darksteel
Reading in set: Fifth Dawn
Reading in set: Champions of Kamigawa
Reading in set: Betrayers of Kamigawa
Reading in set: Saviors of Kamigawa
Reading in set: Ninth Edition
Reading in set: Ravnica: City of Guilds
Reading in set: Guildpact
Reading in set: Dissension
Reading in set: Coldsnap
Reading in set: Time Spiral
Reading in set: Time Spiral "Timeshifted"
Reading in set: Planar Chaos
Reading in set: Future Sight
Reading in set: Tenth Edition
Reading in set: Lorwyn
Reading in set: Morningtide
Reading in set: Shadowmoor
Reading in set: Eventide
Reading in set: Shards of Alara
Reading in set: Conflux
Reading in set: Alara Reborn
Reading in set: Magic 2010
Reading in set: Zendikar
Reading in set: Worldwake
Reading in set: Rise of the Eldrazi
Reading in set: Magic 2011
Reading in set: Scars of Mirrodin
Reading in set: Mirrodin Besieged
Reading in set: New Phyrexia
Reading in set: Magic 2012
Reading in set: Inni

In [47]:
# store to disk

cards_no_nulls.to_pickle('data/all_cards_modern_no_name_hardmode.pkl')